torch/quantization/_adaround.py

```diff
@@ -0,0 +1,260 @@
+    '''
+    return dict(model.named_modules())[name]
+
+def parent_child_names(name):
+    '''Splits full name of submodule into parent submodule's full name and submodule's name
+    '''
+    split_name = name.rsplit('.', 1)
+    if len(split_name) == 1:
+        return '', split_name[0]
+    else:
+        return split_name[0], split_name[1]
+
+class adaround(FakeQuantize):
+    def __init__(self, *args, **keywords):
+        super(adaround, self).__init__(*args, **keywords)
+        self.continous_V = None
+
+    def forward(self, X):
+        if self.continous_V is None:
+            # small initial values are intended so no bias is introduced on initialization or calibration
+            self.continous_V = torch.nn.Parameter(torch.ones(X.size()) / 10000)
+
+        if self.observer_enabled[0] == 1:
+            self.activation_post_process(X.detach())
+            _scale, _zero_point = self.calculate_qparams()
+            _scale, _zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            self.scale = _scale
+            self.zero_point = _zero_point
+
+        if self.fake_quant_enabled[0] == 1:
+            X = adaround_rounding(self, X)
+            if self.qscheme == torch.per_channel_symmetric or self.qscheme == torch.per_channel_affine:
+                X = torch.fake_quantize_per_channel_affine(X, self.scale, self.zero_point,
+                                                           self.ch_axis, self.quant_min, self.quant_max)
+            else:
+                X = torch.fake_quantize_per_tensor_affine(X, float(self.scale),
+                                                          int(self.zero_point), self.quant_min,
+                                                          self.quant_max)
+        return X
+
+    def randomize(self):
+        # uniform distribution of vals
+        init.kaiming_uniform_(self.continous_V, a=math.sqrt(5))
+
+
+class OuputWrapper(nn.Module):
+    def __init__(self, model):
+        super(OuputWrapper, self).__init__()
+        self.wrapped_module = model
+
+        self.float_output = None
+        self.quantized_output = None
+
+        self.quant = QuantStub()
+        self.dequant = DeQuantStub()
+
+    def forward(self, x):
+        x = self.quant(x)
+        self.wrapped_module.activation_post_process.disable_fake_quant()
+        self.wrapped_module.weight_fake_quant.disable_fake_quant()
+        self.float_output = self.wrapped_module(x).detach()
+
+        self.wrapped_module.activation_post_process.enable_fake_quant()
+        self.wrapped_module.weight_fake_quant.enable_fake_quant()
+        self.quantized_output = self.wrapped_module(x)
+
+        print("norm in forward outputwrapper: ", torch.norm(self.float_output - self.quantized_output))
+        return self.dequant(self.quantized_output)
+
+
+araround_fake_quant = adaround.with_args(observer=HistogramObserver, quant_min=-128, quant_max=127,
```

Comment:
btw, using HistogramObserver might be the reason that it is slow, you can try MinMax to make it faster when you run tests.